ssl/record/ssl3_record.c

```diff
@@ -2030,3 +2030,28 @@ int dtls1_get_record(SSL *s)
     return 1;
 
 }
+
+int dtls_buffer_listen_record(SSL *s, size_t len, unsigned char *seq, size_t off)
+{
+    SSL3_RECORD *rr;
+
+    rr = RECORD_LAYER_get_rrec(&s->rlayer);
+    memset(rr, 0, sizeof(SSL3_RECORD));
+
+    rr->length = len;
+    rr->type = SSL3_RT_HANDSHAKE;
+    memcpy(rr->seq_num, seq, sizeof(rr->seq_num));
+    rr->off = off;
+
+    s->rlayer.packet = RECORD_LAYER_get_rbuf(&s->rlayer)->buf;
```

Comment:
This function (dtls_buffer_listen_record()) is the part of this PR that I am least confident about.  Everything else looks good, but I want to double-check the memory semantics here.
We're setting the rlayer.packet to point to the rbuf's storage, and then calling dtls1_buffer_record().  This could be problematic if the "buffering" was just keeping a pointer reference, as the rbuf could get reallocated, the contents overwritten, etc.  But it looks like dtls1_buffer_record() is internally giong to do a memcpy to the DTLS1_RECORD_DATA's local storage and clear out the ```s->rlayer.packet``` field, avoiding the risk of any memory reference extending past the scope of this function.
So I tentatively think it all checks out, but please confirm that I'm understanding properly before merging.

Reply:
The SSL3_BUFFER structure contains meta data about the data buffer , as well as a pointer to the malloc'd data buffer itself. "packet" points into the data buffer. `dtls_buffer_record` takes a copy of the entire SSL3_BUFFER structure - which includes the pointer to the data buffer - and then resets the whole thing to 0. It also takes a copy of "packet" - with the original also zero'd afterwards. So the original data buffer remain "active", and allocated (it isn't copied anywhere) but there are only two things that point into it: the copied SSL3_BUFFER structure that is pushed onto the queue, as well as the copy of "packet" which is also pushed on the queue as part of the same item. Whilst on the queue there is no possibility of the data buffer being reallocated.

Hopefully that helps?

Reply:
It does, thanks.